src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala

```diff
@@ -0,0 +1,242 @@
+    dfFromSample.write.json(sampleJsonDataLocation)
+
+    df = spark.read.parquet(sampleParquetDataLocation)
+    hyperspace.createIndex(df, indexConfig1)
+    copyFirstFile(df)
+
+    df = spark.read.json(sampleJsonDataLocation)
+    hyperspace.createIndex(df, indexConfig2)
+    copyFirstFile(df)
+  }
+
+  after {
+    spark.disableHyperspace()
+  }
+
+  override def afterAll(): Unit = {
+    super.afterAll()
+    FileUtils.delete(new Path(sampleParquetDataLocation))
+    FileUtils.delete(new Path(sampleJsonDataLocation))
+  }
+
+  test("Append-only: filter index & parquet format, " +
+    "index relation should include appended file paths") {
+    df = spark.read.parquet(sampleParquetDataLocation)
+    val query = df.filter(df("clicks") <= 2000).select(df("query"))
+
+    withSQLConf("spark.hyperspace.index.hybridscan.enabled" -> "false") {
+      val transformed = FilterIndexRule(query.queryExecution.optimizedPlan)
+      assert(transformed.equals(query.queryExecution.optimizedPlan), "No plan transformation.")
+    }
+
+    withSQLConf("spark.hyperspace.index.hybridscan.enabled" -> "true") {
+      val planWithHybridScan = FilterIndexRule(query.queryExecution.optimizedPlan)
+      assert(
+        !planWithHybridScan.equals(query.queryExecution.optimizedPlan),
+        "Plan should be transformed.")
+
+      // check appended file is added to relation node or not
+      val nodes = planWithHybridScan collect {
+        case p @ LogicalRelation(fsRelation: HadoopFsRelation, _, _, _) =>
+          assert(fsRelation.location.inputFiles.count(_.contains(".copy")) === 1)
+          assert(fsRelation.location.inputFiles.count(_.contains("index1")) === 4)
+          p
+      }
+      assert(nodes.length === 1)
+
+      spark.enableHyperspace()
+      val query2 = df.filter(df("clicks") <= 2000).select(df("query"))
+
+      checkAnswer(query2, query)
+    }
+  }
+
+  test("Append-only: join index, appended data should be shuffled with indexed columns " +
+    "and merged by BucketUnion") {
+    df = spark.read.parquet(sampleParquetDataLocation)
+    val query = df.filter(df("clicks") >= 2000).select(df("clicks"), df("query"))
+    val query2 = df.filter(df("clicks") <= 4000).select(df("clicks"), df("query"))
+    val join = query.join(query2, "clicks")
+
+    withSQLConf("spark.sql.autoBroadcastJoinThreshold" -> "-1") {
+      withSQLConf("spark.hyperspace.index.hybridscan.enabled" -> "false") {
+        val transformed = JoinIndexRule(join.queryExecution.optimizedPlan)
+        assert(transformed.equals(join.queryExecution.optimizedPlan), "No plan transformation.")
+      }
+
+      withSQLConf("spark.hyperspace.index.hybridscan.enabled" -> "true") {
+        val planWithHybridScan = JoinIndexRule(join.queryExecution.optimizedPlan)
+        assert(
+          !planWithHybridScan.equals(join.queryExecution.optimizedPlan),
+          "Plan should be transformed.")
+
+        // check appended file is added to relation node or not
```

Comment:
Comments should start with upper case and end with "." (if it's a sentence). Can you go thru your comments?